flink-kubernetes/src/main/java/org/apache/flink/kubernetes/kubeclient/resources/KubernetesSharedInformer.java

```diff
@@ -0,0 +1,244 @@
+                        if (removeHandler) {
+                            handlers.remove(resourceKey);
+                        }
+                    },
+                    executorService);
+            return () -> closeFuture.complete(null);
+        }
+
+        private Optional<EventHandler> findHandler(T obj) {
+            final String resourceKey = getResourceKey(obj.getMetadata().getName());
+            return Optional.ofNullable(handlers.get(resourceKey));
+        }
+    }
+
+    private class EventHandler {
+        private final String resourceKey;
+        private final Map<String, WatchCallback<R>> callbacks = new HashMap<>();
+
+        private T resource;
+
+        private EventHandler(String resourceKey) {
+            this.resourceKey = resourceKey;
+        }
+
+        private void addWatch(String id, WatchCallback<R> callback) {
+            log.info("Starting to watch for {}, watching id:{}", resourceKey, id);
+            callbacks.put(id, callback);
+            if (resource != null) {
+                callback.run(h -> h.onAdded(wrapEvent(resource)));
+            }
+        }
+
+        private boolean removeWatch(String id) {
+            callbacks.remove(id);
+            log.info("Stopped to watch for {}, watching id:{}", resourceKey, id);
+            return callbacks.isEmpty();
+        }
+
+        private void onAdded(T obj) {
+            resource = obj;
+            this.callbacks.forEach((id, callback) -> callback.run(h -> h.onAdded(wrapEvent(obj))));
+        }
+
+        private void onModified(T obj) {
+            resource = obj;
+            this.callbacks.forEach(
+                    (id, callback) -> callback.run(h -> h.onModified(wrapEvent(obj))));
+        }
+
+        private void onDeleted(T obj) {
+            resource = null;
+            this.callbacks.forEach(
+                    (id, callback) -> callback.run(h -> h.onDeleted(wrapEvent(obj))));
+        }
+
+        private List<R> wrapEvent(T obj) {
+            return Collections.singletonList(eventWrapper.apply(obj));
+        }
+
+        private void handleExceptionEvent(Exception e) {
+            this.callbacks.forEach((id, callback) -> callback.run(h -> h.handleError(e)));
+        }
+    }
+
+    private static final class WatchCallback<T> {
+        private final WatchCallbackHandler<T> handler;
+        private final ExecutorService executorService;
+
+        private WatchCallback(
+                WatchCallbackHandler<T> handler, @Nullable ExecutorService executorService) {
+            this.handler = handler;
+            this.executorService = executorService;
+        }
+
+        private void run(Consumer<WatchCallbackHandler<T>> handlerConsumer) {
+            if (this.executorService == null) {
+                handlerConsumer.accept(handler);
+            } else {
+                this.executorService.submit(() -> handlerConsumer.accept(handler));
```

Comment:
I keep thinking about the ordering guarantee you've brought up and this may be a right place to handle it. Each callback can have it's own "callbackLock", which we could use to make things completely deterministic and still leverage what executor service, so we don't block the watcher thread 🤔 
```suggestion
                this.executorService.submit(() -> {
                    synchronized(callbackLock) {
                        handlerConsumer.accept(handler);
                    }
                );
```

Reply:
great idea, but a simple synchronized is not enough, i use a `ReentrantLock(true)` for it.

Reply:
yeah, fair lock makes sense here 👍 nice catch